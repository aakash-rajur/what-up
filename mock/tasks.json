[
  {
    "id": "1533233094860",
    "description": "task1",
    "status": "COMPLETED",
    "created": "1533233094860",
    "updated": "1533233094860"
  },
  {
    "id": "1533233143045",
    "description": "task2",
    "status": "CREATED",
    "created": "1533233143045",
    "updated": "1533233143045"
  }
]
