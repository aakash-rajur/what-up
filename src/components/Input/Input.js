import React, {Component} from 'react';
import PropTypes from 'prop-types';

class Input extends Component {
    static propTypes = {
        onValueReturn: PropTypes.func
    };

    constructor(props) {
        super(props);
        this.onReturn = this.onReturn.bind(this);
    }

    render() {
        let inputProps = {...this.props};
        delete inputProps.onValueReturn;
        return (
            <input {...inputProps} onKeyUp={this.onReturn}/>
        )
    }

    onReturn(e) {
        let {
            keyCode,
            target: {value}
        } = e, {
            onKeyUp,
            onValueReturn
        } = this.props;
        onKeyUp && onKeyUp(e);
        keyCode === 13 && onValueReturn && onValueReturn(value)
    }
}

export default Input;
