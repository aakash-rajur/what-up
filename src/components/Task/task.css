@import "../../utils/variables.css";

.task {
    min-height: 70px;
    border: none;
    padding: 0.9em 1em 0.4em 1em;
    font-size: 1.5em;
    font-weight: 100;
    border-radius: 4px;
    color: gray;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    transition: all 0.3s cubic-bezier(.25, .8, .25, 1);
    background: white;
    z-index: 0;

    display: grid;
    grid-template: 1fr/2em 1fr auto;
    align-items: center;
}

.status {
    border: 1px solid lightgray;
    border-radius: 50%;
    position: relative;
}

.status:after {
    content: "✔";
    position: absolute;
    left: 0;
    top: 0;
    font-size: 0.6em;
    width: var(--icon-size);
    height: var(--icon-size);
    display: flex;
    align-items: center;
    justify-content: center;
    color: limegreen;
    opacity: 0;
    transition: all 0.3s;
}

.cancel {
    visibility: hidden;
}

.task:hover:not([data-status="CANCELLED"]) > .cancel {
    visibility: initial;
}

.task[data-status="COMPLETED"] > .status:after {
    opacity: 1;
}

.task[data-status="COMPLETED"] > .description {
    text-decoration: line-through;
    text-decoration-color: var(--accent-opaque);
    opacity: 0.5;
}

.task[data-status="CANCELLED"] > .status:after {
    opacity: 1;
    content: "✗";
    color: gray;
}

.task[data-status="CANCELLED"] > .description {
    text-decoration: line-through;
    opacity: 0.5;
}

.editable, .description {
    font-weight: 100;
    font-size: 1em;
}

.editable {
    border: none;
    outline: none;
    color: gray;
}

.description {
    transition: all 0.3s;
    cursor: pointer;
}

.updated {
    font-size: 0.4em;
    grid-column: span 3;
    font-weight: 100;
    justify-self: end;
    color: lightgray;
}

@media (max-width: 600px) {
    .task {
        font-size: 1.2em;
    }
}
