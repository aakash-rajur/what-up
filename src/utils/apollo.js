import {InMemoryCache} from 'apollo-cache-inmemory';
import {ApolloClient} from 'apollo-client';
import {ApolloLink, split} from 'apollo-link';
import {onError} from 'apollo-link-error';
import {HttpLink} from 'apollo-link-http';
import {WebSocketLink} from 'apollo-link-ws';
import {getMainDefinition} from "apollo-utilities";
import gql from 'graphql-tag';
import {compose, graphql} from 'react-apollo';
import {
	API_URL,
	SESSION_CHANGE,
	TASK_ALL,
	TASK_CANCELLED,
	TASK_COMPLETED,
	TASK_CREATED,
	TASKS_CHANGED,
	WS_URL
} from "./constants";

const wsLink = new WebSocketLink({
	uri: WS_URL,
	options: {
		reconnect: true
	}
});

export default new ApolloClient({
	link: ApolloLink.from([
		onError(({graphQLErrors, networkError}) => {
			if (graphQLErrors)
				graphQLErrors.map(({message, locations, path}) =>
					console.log(
						`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`,
					),
				);
			if (networkError) console.log(`[Network error]: ${JSON.stringify(networkError)}`);
		}),
		split(({query}) => {
				const {kind, operation} = getMainDefinition(query);
				return kind === 'OperationDefinition' && operation === 'subscription';
			},
			wsLink,
			new HttpLink({
				uri: API_URL,
				credentials: 'include'
			}))
	]),
	cache: new InMemoryCache()
});

export const FETCH_TASKS = gql`
    query fetchTasks($filter: String!, $timestamp: String){
        tasks(filter: $filter, timestamp: $timestamp) {
            id,
            description,
            status,
            created,
            updated
        }
    }
`;

export const CANCEL_TASK = gql`
    mutation removeTask($id: String!) {
        remove(id: $id)
    }
`;

export const EDIT_TASK = gql`
    mutation editTask($id: String!, $description: String!) {
        edit(id: $id, description: $description)
    }
`;

export const UPDATE_TASK = gql`
    mutation updateTask($id: String!, $status: String!) {
        update(id: $id, status: $status)
    }
`;

export const UPDATE_ALL_TASKS = gql`
    mutation updateAll($filter: String!, $status: String!) {
        updateAll(filter: $filter,status: $status)
    }
`;

export const ADD_TASK = gql`
    mutation addTask($description: String!) {
        add(description: $description)
    }
`;

export const TASKS_UPDATED = gql`
    subscription onTasksChanged {
        TASKS_CHANGED {
            timestamp,
            CREATED,
            COMPLETED,
            CANCELLED,
            ALL
        }
    }
`;

export const SESSION_CHANGED = gql`
    subscription onSessionChanged {
        SESSION_CHANGE {
            additional
        }
    }
`;

export const withTaskMutations = compose(...[{
		mutation: CANCEL_TASK,
		prefix: 'cancel',
		args: ({id}) => ({id})
	}, {
		mutation: EDIT_TASK,
		prefix: 'edit',
		args: () => ({}),
		dynamic: true
	}, {
		mutation: UPDATE_TASK,
		prefix: 'update',
		args: ({id, status}) => ({
			id, status: status === TASK_CREATED ?
				TASK_COMPLETED : TASK_CREATED
		})
	}].map(({mutation, prefix, args, dynamic}) => graphql(mutation, {
		props: ({mutate}, result) => {
			return ({
				[`${prefix}Task`]: dynamic ? props => mutate({variables: {...props}}) : mutate,
				[`${prefix}Result`]: result ? result.data : null
			})
		},
		options: props => ({variables: args(props)})
	}))
);

export const withQueryTasks = graphql(FETCH_TASKS, {
	props: ({data = {}}, ...rest) => ({...data, ...rest})
});

export const withNewTaskAddition = graphql(ADD_TASK, {
	props: ({mutate}, result) => ({addNewTask: mutate, result}),
	options: ({newTask, onDone}) => ({
		variables: {description: newTask},
		onCompleted: onDone,
		onError: onDone
	})
});

export const withUpdateAll = graphql(UPDATE_ALL_TASKS, {
	props: ({mutate, result}) => ({updateAll: mutate, result}),
	options: ({filter, nextStatus}) => ({variables: {filter, status: nextStatus}})
});

const defaultTaskUpdated = {
	[TASKS_CHANGED]: {
		timestamp: "0",
		[TASK_ALL]: 0,
		[TASK_CANCELLED]: 0,
		[TASK_COMPLETED]: 0,
		[TASK_CREATED]: 0
	}
};

const defaultSession = {
	[SESSION_CHANGE]: null
};

export const withSessionAndTaskSubscription = compose(
	graphql(SESSION_CHANGED, {
		props: ({data: {[SESSION_CHANGE]: session} = defaultSession}) => ({
			session: session && JSON.parse(session.additional)
		}),
		options: () => ({shouldResubscribe: true})
	}),
	graphql(TASKS_UPDATED, {
		props: ({data: {[TASKS_CHANGED]: stats} = defaultTaskUpdated}) => ({...stats}),
		options: () => ({shouldResubscribe: true})
	}),
);
