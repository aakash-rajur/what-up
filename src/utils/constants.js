export const API_URL = process.env.REACT_APP_API_URL;
export const WS_URL = process.env.REACT_APP_SOCKET_URL;
export const TASKS_CHANGED = process.env.REACT_APP_TASKS_CHANGED;
export const TASK_CREATED = process.env.REACT_APP_TASK_CREATED;
export const TASK_COMPLETED = process.env.REACT_APP_TASK_COMPLETED;
export const TASK_CANCELLED = process.env.REACT_APP_TASK_CANCELLED;
export const TASK_ALL = process.env.REACT_APP_TASK_ALL;
export const SESSION_CHANGE = process.env.REACT_APP_SESSION_CHANGE;
export const TASK_TITLE = 'Task description\nDouble click to edit';
