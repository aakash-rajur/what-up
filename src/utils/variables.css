:root {
    --accent-color: rgba(175, 47, 47, 0.15);
    --accent-opaque: #cc9a9a;
    --app-width: 550px;
    --icon-size: 24px;
}

@media (max-width: 600px) {
    :root {
        --icon-size: 20px;
    }
}
