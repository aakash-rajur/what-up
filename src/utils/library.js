import React from "react";
import {Mutation} from 'react-apollo';
import {TASK_ALL, TASK_CANCELLED, TASK_COMPLETED, TASK_CREATED} from "./constants";

export function promiseSetState(context) {
	return newState => new Promise(resolve =>
		context.setState(newState, resolve)
	);
}

export function wait(delay, cb, ...args) {
	let timeout = null,
		promise = new Promise(resolve => {
			timeout = setTimeout(async () =>
				resolve(cb && await cb(...(args || []))),
				delay
			);
		});
	promise.cancel = () => clearTimeout(timeout);
	return promise;
}

export function renderMutations(config) {
	return config.map(({mutation, getVariables, renderComponent}, index) => {
		if (!(renderComponent instanceof Function)) return renderComponent;
		return (
			<Mutation key={index} mutation={mutation} variables={getVariables(mutation)}>
				{renderComponent}
			</Mutation>
		)
	});
}

export function getNextUpdateAllStatus(tasks = []) {
	let task = tasks.find(({status}) => status !== TASK_COMPLETED);
	return task ? TASK_COMPLETED : TASK_CREATED;
}

const DAY_MAP = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
const MONTH_MAP = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];

function padStart(string) {
	return typeof string === 'string' ? string.padStart(2, '0') :
		string.toString().padStart(2, '0');
}

export function getFormattedTimestamp(timestamp) {
	let date = new Date(timestamp);
	// noinspection JSCheckFunctionSignatures
	if (isNaN(date)) return 'Invalid Date';
	let hr = date.getHours(),
		suffix = 'A.M';
	if (hr === 0) {
		hr = 12;
		suffix = 'A.M';
	} else if (hr === 12) {
		suffix = 'P.M';
	} else hr %= 12;
	return `${padStart(hr)}: ${padStart(date.getMinutes())} ${suffix}, ${DAY_MAP[date.getDay()]} ${date.getDate()} ${MONTH_MAP[date.getMonth()]}`
}

export const FILTER_BUTTON_TEMPLATE = [{
	text: 'All',
	title: 'View All Tasks',
	filter: TASK_ALL
}, {
	text: 'Active',
	title: 'View Active Tasks',
	filter: TASK_CREATED
}, {
	text: 'Completed',
	title: 'View Completed Tasks',
	filter: TASK_COMPLETED
}, {
	text: 'Cancelled',
	title: 'View Cancelled Tasks',
	filter: TASK_CANCELLED
}];
