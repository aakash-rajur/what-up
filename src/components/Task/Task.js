import PropTypes from 'prop-types';
import React, {Component} from 'react';
import CANCEL_ICON from '../../assets/close.svg';
import {MutableTask} from "../../utils/apollo";
import {TASK_CREATED, TASK_TITLE} from "../../utils/constants";
import {getFormattedTimestamp} from "../../utils/library";
import Input from "../Input/Input";
import './task.css';

export class Task extends Component {
	static propTypes = {
		description: PropTypes.string,
		status: PropTypes.string,
		id: PropTypes.string,
		updated: PropTypes.string
	};
	
	constructor(props) {
		super(props);
		this.onDescriptionChange = this.onDescriptionChange.bind(this);
		this.onSaveDescription = this.onSaveDescription.bind(this);
		this.state = {
			description: props.description,
			editable: false
		};
	}
	
	render() {
		let {
			status,
			updated,
			updateTask,
			cancelTask
		} = this.props, {
			description,
			editable
		} = this.state;
		return (
			<div className="app-width task" data-status={status}>
				<div key='status' className="icon status"
				     title={status === TASK_CREATED ?
					     'Set Complete' : 'Set Active'}
				     onClick={updateTask}>
					&nbsp;
				</div>
				{editable ? (
					<Input key='description' value={description}
					       className='editable' autoFocus={true}
					       onChange={this.onDescriptionChange}
					       onBlur={this.onEditChange(false)}
					       onValueReturn={this.onSaveDescription}/>
				) : (
					<div className="description" key='description'
					     title={TASK_TITLE} onDoubleClick={this.onEditChange(true)}>
						{description}
					</div>
				)}
				<img key='remove' src={CANCEL_ICON} alt="cancel-task"
				     className="icon cancel" title='Set Cancel'
				     onClick={cancelTask}/>
				<i className="updated">{getFormattedTimestamp(updated)}</i>
			</div>
		)
	}
	
	onEditChange(editable) {
		return () => {
			let {description} = this.props;
			this.setState({editable, description: description.trim()});
		};
	}
	
	onDescriptionChange({target: {value}}) {
		this.setState({description: value});
	}
	
	async onSaveDescription() {
		let {id, editTask} = this.props,
			{description} = this.state;
		await editTask({id, description});
		this.setState({editable: false});
	}
}


export default MutableTask(Task);
