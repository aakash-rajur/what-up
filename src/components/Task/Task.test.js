import React from 'react';
import tasks from '../../../mock/tasks'
import CANCEL_ICON from "../../assets/close.svg";
import {TASK_TITLE} from "../../utils/constants";
import {getFormattedTimestamp} from "../../utils/library";
import Input from "../Input/Input";
import {Task} from './Task';

describe('Task component non editable description', () => {
	let updateTask = stub(),
		cancelTask = stub(),
		editTask = stub(),
		props = {
			...tasks[0],
			updateTask,
			cancelTask,
			editTask,
		}, component = shallow(<Task {...props}/>);
	
	it('should render', () => {
		expect(component.find('div.app-width.task').exists())
			.toBe(true);
	});
	
	it('dom integrity', checkDOM(component, [{
		selector: 'div.app-width.task',
		props: {
			'data-status': tasks[0].status
		},
		children: [{
			selector: 'div',
			length: 3
		}, {
			selector: 'img',
			length: 1
		}, {
			selector: 'i',
			length: 1
		}]
	}, {
		selector: 'div.icon.status',
		props: {
			title: 'Set Active',
			onClick: updateTask
		},
		text: '\u00a0'
	}, {
		selector: 'div.description',
		props: {
			title: TASK_TITLE
		},
		text: tasks[0].description
	}, {
		selector: 'img.icon.cancel',
		props: {
			src: CANCEL_ICON,
			alt: 'cancel-task',
			title: 'Set Cancel',
			onClick: cancelTask
		}
	}, {
		selector: 'i.updated',
		text: getFormattedTimestamp(tasks[0].updated)
	}]));
	
	it('state integrity', () => [{
			prop: 'description',
			value: tasks[0].description
		}, {
			prop: 'editable',
			value: false
		}].forEach(({prop, value}) =>
			expect(component.state(prop))
				.toBe(value)
		)
	);
	
	[{
		test: 'status',
		selector: 'div.icon.status',
		callback: 'updateTask'
	}, {
		test: 'cancel',
		selector: 'img.icon.cancel',
		callback: 'cancelTask'
	}].forEach(({test, selector, callback}) =>
		it(`should invoke ${callback} on ${test} click`, () => {
			component.find(selector)
				.simulate('click');
			let func = props[callback];
			expect(func.called).toBe(true);
			expect(func.callCount).toBe(1);
		})
	);
	
	matchSnapshot(component);
});

describe('Task component editable description', () => {
	let updateTask = stub(),
		cancelTask = stub(),
		editTask = stub(),
		props = {
			...tasks[0],
			updateTask,
			cancelTask,
			editTask,
		}, component = shallow(<Task {...props}/>);
	
	component.find('div.description')
		.simulate('doubleClick');
	
	it('should render', () =>
		expect(component.find('div.app-width.task').exists())
			.toBe(true)
	);
	
	it('state integrity after', () => [{
			prop: 'description',
			value: tasks[0].description
		}, {
			prop: 'editable',
			value: true
		}].forEach(({prop, value}) =>
			expect(component.state(prop))
				.toBe(value)
		)
	);
	
	it('dom integrity', checkDOM(component, [{
		selector: 'div.app-width.task',
		props: {
			'data-status': tasks[0].status
		},
		children: [{
			selector: 'div',
			length: 2
		}, {
			selector: 'img',
			length: 1
		}, {
			selector: 'i',
			length: 1
		}]
	}, {
		selector: 'div.icon.status',
		props: {
			title: 'Set Active',
			onClick: updateTask
		},
		text: '\u00a0'
	}, {
		selector: Input,
		props: {
			value: tasks[0].description,
			className: 'editable',
			autoFocus: true
		}
	}, {
		selector: 'img.icon.cancel',
		props: {
			src: CANCEL_ICON,
			alt: 'cancel-task',
			title: 'Set Cancel',
			onClick: cancelTask
		}
	}, {
		selector: 'i.updated',
		text: getFormattedTimestamp(tasks[0].updated)
	}]));
	
	it('check callbacks', checkEvents(component, [{
		selector: 'div.icon.status',
		event: 'click',
		callback: updateTask,
		count: 1
	}, {
		selector: 'img.icon.cancel',
		event: 'click',
		callback: cancelTask,
		count: 1
	}, {
		selector: Input,
		event: 'onValueReturn',
		custom: true,
		callback: editTask,
		cArgs: [{
			id: tasks[0].id,
			description: tasks[0].description
		}]
	}]));
	
	/**
	 * snapshot is being checked before setState
	 */
	setImmediate(matchSnapshot, 100, component);
});
