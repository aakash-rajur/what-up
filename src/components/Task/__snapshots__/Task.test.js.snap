// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task component non editable description match snapshot 1`] = `
<div
  className="app-width task"
  data-status="COMPLETED"
>
  <div
    className="icon status"
    key="status"
    onClick={[Function]}
    title="Set Active"
  >
    Â 
  </div>
  <div
    className="description"
    key="description"
    onDoubleClick={[Function]}
    title="Task description
  Double click to edit"
  >
    task1
  </div>
  <img
    alt="cancel-task"
    className="icon cancel"
    key="remove"
    onClick={[Function]}
    src="close.svg"
    title="Set Cancel"
  />
  <i
    className="updated"
  >
    11: 34 A.M, Thu 2 Aug
  </i>
</div>
`;
